crates/data-generator/src/blocks/json.rs

```diff
@@ -0,0 +1,194 @@
+use std::collections::BTreeMap;
+
+use serde::Deserialize;
+use thiserror::Error;
+
+use super::model::{Block, BlockProperty, BlockPropertyVariant, Id, State, StateProperty};
+
+#[derive(Debug, Deserialize)]
+struct SerdeId(u32);
+
+#[derive(Debug, Deserialize)]
+struct SerdeBlock {
+    #[serde(default)]
+    properties: BTreeMap<String, Vec<String>>,
+    states: Vec<SerdeState>,
+}
+
+#[derive(Debug, Deserialize)]
+struct SerdeState {
+    id: SerdeId,
+    #[serde(default)]
+    default: bool,
+    #[serde(default)]
+    properties: BTreeMap<String, String>,
+}
+
+#[derive(Debug, Error)]
+pub enum ParseError {
+    #[error("malformed JSON data: {0}")]
+    Json(#[from] serde_json::Error),
+}
+
+pub(super) fn parse_block_registry(json_data: String) -> Result<Vec<Block>, ParseError> {
+    let registry: BTreeMap<String, SerdeBlock> = serde_json::from_str(&json_data)?;
+
+    fn block_property(
+        block_name_pc: String,
+        prop_name_sc: String,
+        prop_variants: Vec<String>,
+    ) -> BlockProperty {
+        let prop_name_sc = defuse_property_name(&prop_name_sc);
+        let prop_name_pc = transform_to_pascal_case(&prop_name_sc);
+
+        let all_bools = all_bools(&prop_variants);
+        let all_nums = !all_bools && all_nums(&prop_variants);
+
+        let prop_variants = prop_variants
+            .iter()
+            .map(|prop_variant| {
+                let var_name = transform_to_pascal_case(&prop_variant);
+                let var_name = defuse_variant_name(var_name);
+
+                match (all_bools, all_nums) {
+                    (true, false) => {
+                        let value = prop_variant.parse().expect(&prop_variant);
+                        BlockPropertyVariant::Bool(value, var_name)
+                    }
+                    (false, true) => {
+                        let ord = prop_variant.parse().unwrap();
+                        BlockPropertyVariant::Numeric(ord, var_name)
+                    }
+                    (_, _) => BlockPropertyVariant::Regular(var_name),
+                }
```

Comment:
Replace with a usual if-else-if ladder, or instead of `bool`s store a `fn(var_name, prop_variant) -> BlockPropertyVariant`, and then call it from here regardless of which variant it generates.